### PR TITLE
refactor(iot-service): added logging in service client and improved r…

### PR DESCRIPTION
**Pull request by davilu**
…eading inputstream

 

# Checklist
- [ ] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-java/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [ ] This pull-request is submitted against the `master` branch. 

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java [hunk 1]
```diff
@@ -178,6 +181,7 @@ private synchronized MethodResult invokeMethod(URL url, String methodName, Long
         MethodParser methodParserResponse = new MethodParser();
         methodParserResponse.fromJson(new String(response.getBody(), StandardCharsets.UTF_8));
 
+        log.debug("After invoke: url={}, methodName={}, connectTimeout={} seconds, responseTimeout={} seconds", url, methodName, connectTimeoutInSeconds, responseTimeoutInSeconds);
```
Comment (timtay-microsoft) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java]: >log [](start = 8, length = 3)

Is there a good way to correlate this back to the log above? What if service client invokes the same method on the same device twice from two different threads

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/http/HttpRequest.java [hunk 2]
```diff
@@ -36,6 +38,7 @@ public HttpRequest(URL url, HttpMethod method, byte[] body) throws IOException
         // Codes_SRS_SERVICE_SDK_JAVA_HTTPREQUEST_12_003: [The function shall use the given HTTPS method (i.e. GET) as the request method.]
         // Codes_SRS_SERVICE_SDK_JAVA_HTTPREQUEST_12_004: [If an IOException occurs in setting up the HTTPS connection, the function shall throw an IOException.]
         this.connection = new HttpConnection(url, method);
+        log.debug("Created connection: url={}, method={}, connection={}", url, method, connection);
```
Comment (timtay-microsoft) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/http/HttpRequest.java]: >connection [](start = 27, length = 10)

created request, not connection. connection is created during send()

**Comment by timtay-microsoft** [comment]
>             headerFields = this.connection.getResponseHeaders();

Is it worth logging these, too?

---
Refers to: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/http/HttpRequest.java:84 in 6d86aee. [](commit_id = 6d86aeeab56ce8a4aa7b14049db2724cf2b1d87e, deletion_comment = False)

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/http/HttpRequest.java [hunk 3]
```diff
@@ -51,23 +54,27 @@ public HttpRequest(URL url, HttpMethod method, byte[] body) throws IOException
...
             responseStatus = this.connection.getResponseStatus();
+            log.debug("Connection {} status={}.", connection, responseStatus);
             headerFields = this.connection.getResponseHeaders();
             responseBody = this.connection.readInput();
         }
         // Can be caused either by an unsuccessful
         // connection or by a bad status code.
         catch (IOException e)
         {
+            log.debug("Connection {} failed.", connection, e);
```
Comment (timtay-microsoft) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/http/HttpRequest.java]: >connection [](start = 47, length = 10)

how does this object get logged?

Comment (davilu) [reply service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/http/HttpRequest.java]: class name + hash

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/http/HttpRequest.java (hunk 2 above)
Comment (davilu) [reply service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/http/HttpRequest.java]: `connection` object is created on line 40. and `connection` will be a keyword to track this call.

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java (hunk 1 above)
Comment (davilu) [reply service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java]: Good point. Added `requestId` in the log.

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java [hunk 4]
```diff
@@ -64,7 +64,7 @@ public static DeviceMethod createFromConnectionString(String connectionString) t
      * @throws IotHubException This exception is thrown if the response verification failed
      * @throws IOException This exception is thrown if the IO operation failed
      */
-    public synchronized MethodResult invoke(String deviceId, String methodName, Long responseTimeoutInSeconds, Long connectTimeoutInSeconds, Object payload) throws IotHubException, IOException
+    public MethodResult invoke(String deviceId, String methodName, Long responseTimeoutInSeconds, Long connectTimeoutInSeconds, Object payload) throws IotHubException, IOException
```
Comment (prmathur-microsoft) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java]: Is this a breaking change?

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java [hunk 5]
```diff
@@ -97,7 +97,7 @@ public synchronized MethodResult invoke(String deviceId, String methodName, Long
      * @throws IotHubException This exception is thrown if the response verification failed
      * @throws IOException This exception is thrown if the IO operation failed
      */
-    public synchronized MethodResult invoke(String deviceId, String moduleId, String methodName, Long responseTimeoutInSeconds, Long connectTimeoutInSeconds, Object payload) throws IotHubException, IOException
+    public MethodResult invoke(String deviceId, String moduleId, String methodName, Long responseTimeoutInSeconds, Long connectTimeoutInSeconds, Object payload) throws IotHubException, IOException
```
Comment (prmathur-microsoft) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java]: Isn't this a breaking change?

File: service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java [hunk 6]
```diff
@@ -135,7 +135,7 @@ public synchronized MethodResult invoke(String deviceId, String moduleId, String
      * @throws IotHubException This exception is thrown if the response verification failed
      * @throws IOException This exception is thrown if the IO operation failed
      */
-    private synchronized MethodResult invokeMethod(URL url, String methodName, Long responseTimeoutInSeconds, Long connectTimeoutInSeconds, Object payload) throws IotHubException, IOException
+    private MethodResult invokeMethod(URL url, String methodName, Long responseTimeoutInSeconds, Long connectTimeoutInSeconds, Object payload) throws IotHubException, IOException
```
Comment (prmathur-microsoft) [inline service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/devicetwin/DeviceMethod.java]: Same.. Breaking change!

**Pull request closed without merging**